File: conformance/ts/src/recorded-wire.ts
/** Acceptance-only recorded wire; the generated record/follow API is a later lane. */
import {
  at,
  mount,
  WireError,
  type Message,
  type Path,
  type Receiver,
  type Wire,
  type Endpoint,
} from '@nightseam/duplex';
import { createDispatcher, type HandlerRegistry, type WireDispatcher } from '@nightseam/runtime';

class Gate {
  readonly promise: Promise<void>;
  open!: () => void;
  constructor() {
    this.promise = new Promise((resolve) => {
      this.open = resolve;
    });
  }
}

class Queue<T> {
  readonly items: T[] = [];
  private readonly waiters: Array<(value: T) => void> = [];
  put(value: T): void {
    const waiter = this.waiters.shift();
    if (waiter) waiter(value);
    else this.items.push(value);
  }
  take(): Promise<T> {
    return this.items.length
      ? Promise.resolve(this.items.shift()!)
      : new Promise((resolve) => this.waiters.push(resolve));
  }
}

interface Entry {
  path: Path;
  message: Message;
  sequence: number;
}
class Follower {
  readonly live = new Queue<Entry>();
  readonly sent = new Queue<number>();
  readonly paused = new Gate();
  readonly resume = new Gate();
  readonly stop = new Gate();
  stopped = false;
  done: Promise<void> = Promise.resolve();
  end(): void {
    this.stopped = true;
    this.stop.open();
  }
}

class RecordedWire implements Wire {
  private readonly entries: Entry[] = [];
  private readonly followers = new Map<Follower, number>();
  private appending = false;

  // A callback must be able to reenter the store without append exclusion.
  head(): number {
    if (this.appending) throw new Error('application callback ran under append exclusion');
    return this.entries.length;
  }

  send(path: Path, message: Message): void {
    this.appending = true;
    try {
      const entry = { path: [...path], message, sequence: this.entries.length + 1 };
      this.entries.push(entry);
      for (const [follower, bound] of this.followers) {
        if (follower.live.items.length === bound) {
          this.followers.delete(follower);
          // Signal only: the subscriber writer closes its own carrier outside
          // append and off the producer's stack.
          follower.end();
        } else follower.live.put(entry);
      }
    } finally {
      this.appending = false;
    }
  }
  close(): void {
    for (const follower of this.followers.keys()) follower.end();
    this.followers.clear();
  }

  attach(after: number, target: HandlerRegistry, bound: number, pause: boolean): { head: number; follower: Follower } {
    const follower = new Follower();
    // One synchronous critical section, without an await: append cannot occur
    // between capturing this head and registering its bounded handoff.
    this.appending = true;
    const head = this.entries.length;
    const history = this.entries.slice(after, head);
    this.followers.set(follower, bound);
    this.appending = false;
    follower.done = Promise.resolve().then(async () => {
      const send = (entry: Entry): boolean => {
        if (follower.stopped) return false;
        target.send(entry.path, entry.message);
        follower.sent.put(entry.sequence);
        return true;
      };
      try {
        for (let i = 0; i < history.length; i++) {
          if (!send(history[i]!)) return;
          if (i === 0 && pause) {
            follower.paused.open();
            await Promise.race([follower.resume.promise, follower.stop.promise]);
            if (follower.stopped) return;
          }
        }
        while (!follower.stopped) {
          const entry = await Promise.race([follower.live.take(), follower.stop.promise.then(() => undefined)]);
          if (!entry || !send(entry)) return;
        }
      } finally {
        target.close(1008, 'recorded handoff ended');
      }
    });
    return { head, follower };
  }
}

/** Test root providing queued asynchronous delivery, without a transport API. */
class RecordedRoot implements Endpoint {
  private receiver: Receiver | undefined;
  private readonly queue: Array<{ path: Path; message: Message }> = [];
  private scheduled = false;
  private closed = false;

  send(path: Path, message: Message): void {
    if (this.closed) throw new WireError('closed');
    if (this.queue.length === 16) throw new Error('witness output queue full');
    this.queue.push({ path: [...path], message });
    if (this.scheduled) return;
    this.scheduled = true;
    queueMicrotask(() => {
      this.scheduled = false;
      while (!this.closed && this.queue.length) {
        const entry = this.queue.shift()!;
        this.receiver?.message?.([...entry.path], entry.message);
      }
    });
  }
  receive(receiver: Receiver): () => void {
    if (this.receiver) throw new WireError('receiver_exists');
    this.receiver = receiver;
    return () => {
      if (this.receiver === receiver) this.receiver = undefined;
    };
  }
  close(): void {
    this.closed = true;
    this.queue.length = 0;
    this.receiver = undefined;
  }
}

const message = (value: number): Message => ({ frame: { version: 1, kind: 'event', data: value } });
type Wait = <T>(work: Promise<T>) => Promise<T>;

class Presentation {
  readonly root = new RecordedRoot();
  private readonly end = new RecordedRoot();
  readonly values = new Queue<number>();
  readonly closed = new Queue<number>();
  readonly wire: HandlerRegistry;
  readonly rootDispatch: WireDispatcher;
  private readonly owned: Endpoint[] = [];
  private readonly registries: WireDispatcher[] = [];
  private failure?: unknown;

  constructor(store: RecordedWire) {
    const registry = (endpoint: Endpoint) => {
      const dispatcher = createDispatcher(endpoint);
      this.registries.push(dispatcher);
      return dispatcher;
    };
    this.rootDispatch = registry(this.root);
    const end = registry(this.end);
    const out = mount(new Map([['out', end.select(['destination'])]]));
    const inner = mount(new Map([['in', this.rootDispatch.select(['source'])]]));
    const outer = mount(new Map([['outer', inner]]));
    this.owned.push(out, inner, outer);
    const destination = registry(registry(out).select(['out']));
    this.wire = registry(registry(outer).select(['outer', 'in']));
    destination.register(['tick'], {
      message: (path, message) => {
        try {
          if (path.length !== 1 || path[0] !== 'tick') throw new Error('recorded destination received wrong path');
          store.head(); // Real application reentry, outside append exclusion.
          if (message.frame.kind !== 'event' || typeof message.frame.data !== 'number')
            throw new Error('expected event');
          this.values.put(message.frame.data);
        } catch (error) {
          this.failure = error;
          this.values.put(0);
        }
      },
    });
    this.wire.register(['tick'], {
      message: (path, message) => {
        try {
          destination.send(path, message);
        } catch (error) {
          this.failure = error;
          this.values.put(0);
        }
      },
      closed: (code) => {
        store.head();
        this.closed.put(code);
      },
    });
  }

  async collect(wait: Wait): Promise<number[]> {
    // A fence on the same route follows every earlier replay/live delivery.
    this.wire.send(['tick'], message(0));
    const values: number[] = [];
    for (;;) {
      const value = await wait(this.values.take());
      if (this.failure) throw this.failure;
      if (value === 0) return values;
      values.push(value);
    }
  }
  close(): void {
    for (const registry of [...this.registries].reverse()) registry.close();
    for (const endpoint of [...this.owned].reverse()) endpoint.close();
    this.root.close();
    this.end.close();
  }
}

async function sent(wait: Wait, follower: Follower, last: number): Promise<void> {
  while ((await wait(follower.sent.take())) !== last) {
    /* Wait for the writer's acceptance barrier. */
  }
}

async function headCase(wait: Wait, before: boolean): Promise<unknown> {
  const store = new RecordedWire();
  const first = new Presentation(store);
  const late = new Presentation(store);
  const followers: Follower[] = [];
  try {
    const source = at(mount(new Map([['record', store]])), ['record']);
    for (let value = 1; value <= 3; value++) source.send(['tick'], message(value));
    if (before) source.send(['tick'], message(4));
    const attached = store.attach(0, first.wire, 2, true);
    followers.push(attached.follower);
    await wait(attached.follower.paused.promise);
    // Await producer completion while replay remains at its barrier.
    await wait(
      Promise.resolve().then(() => {
        for (let value = before ? 5 : 4; value <= 5; value++) source.send(['tick'], message(value));
      }),
    );
    attached.follower.resume.open();
    await sent(wait, attached.follower, 5);
    source.send(['tick'], message(6));
    await sent(wait, attached.follower, 6);
    const values = await first.collect(wait);
    const second = store.attach(3, late.wire, 2, false);
    followers.push(second.follower);
    await sent(wait, second.follower, 6);
    return {
      cut: before ? 'append_before_head' : 'head_before_append',
      head: attached.head,
      first: values,
      after_three: await late.collect(wait),
      producer_progress: true,
      callbacks_outside_append: true,
    };
  } finally {
    store.close();
    await Promise.all(followers.map((follower) => follower.done));
    first.close();
    late.close();
  }
}

async function stallCase(wait: Wait): Promise<unknown> {
  const store = new RecordedWire();
  const stalled = new Presentation(store);
  const healthy = new Presentation(store);
  const followers: Follower[] = [];
  try {
    for (let value = 1; value <= 3; value++) store.send(['tick'], message(value));
    const slow = store.attach(0, stalled.wire, 2, true).follower;
    followers.push(slow);
    await wait(slow.paused.promise);
    const fast = store.attach(3, healthy.wire, 2, false).follower;
    followers.push(fast);
    for (let value = 4; value <= 5; value++) {
      store.send(['tick'], message(value));
      await sent(wait, fast, value);
    }
    const queued = slow.live.items.length;
    store.send(['tick'], message(6));
    await sent(wait, fast, 6);
    await wait(slow.done);
    const code = await wait(stalled.closed.take());
    let refused = false;
    try {
      stalled.wire.send(['tick'], message(99));
    } catch (error) {
      refused = error instanceof WireError && error.code === 'closed';
    }
    if (!refused) throw new Error('stalled carrier accepted after close');
    const underneath = new Queue<number>();
    stalled.rootDispatch.register(['probe'], {
      message: (_path, message) => {
        if (message.frame.kind === 'event') underneath.put(message.frame.data as number);
      },
    });
    stalled.root.send(['probe'], message(99));
    const probe = await wait(underneath.take());
    store.send(['tick'], message(7));
    await sent(wait, fast, 7);
    return {
      bound: 2,
      queued_at_bound: queued,
      closed: 1 + stalled.closed.items.length,
      close_code: code,
      healthy: await healthy.collect(wait),
      underneath: [probe],
      head: store.head(),
      producer_progress: true,
    };
  } finally {
    store.close();
    await Promise.all(followers.map((follower) => follower.done));
    stalled.close();
    healthy.close();
  }
}

export async function recordedWireWitness(withinMs: number): Promise<unknown> {
  let timer: NodeJS.Timeout | undefined;
  const deadline = new Promise<never>((_, reject) => {
    timer = setTimeout(() => reject(new Error('recorded wire witness deadline')), withinMs);
  });
  const wait: Wait = (work) => Promise.race([work, deadline]);
  try {
    return { cases: [await headCase(wait, false), await headCase(wait, true)], stalled: await stallCase(wait) };
  } finally {
    clearTimeout(timer);
  }
}

File: conformance/python/recorded_wire.py
"""Acceptance-only consumer composition; record/follow is not a runtime API."""

import asyncio
from collections import deque
from dataclasses import dataclass

from bitwire import Message, Path, Receiver, Wire
from nightseam.duplex.wire import WireError, at, mount
from nightseam.runtime import Dispatcher


class _Queue:
    def __init__(self):
        self.items = deque()
        self.waiters = deque()

    def put(self, value):
        while self.waiters:
            waiter = self.waiters.popleft()
            if not waiter.done():
                waiter.set_result(value)
                return
        self.items.append(value)

    async def take(self):
        if self.items:
            return self.items.popleft()
        waiter = asyncio.get_running_loop().create_future()
        self.waiters.append(waiter)
        try:
            return await waiter
        finally:
            if waiter in self.waiters:
                self.waiters.remove(waiter)


@dataclass(frozen=True)
class _Entry:
    path: Path
    message: Message
    sequence: int


class _Follower:
    def __init__(self):
        self.live = _Queue()
        self.sent = _Queue()
        self.paused = asyncio.Event()
        self.resume = asyncio.Event()
        self.stop = asyncio.Event()
        self.stopped = False
        self.done = None

    def end(self):
        self.stopped = True
        self.stop.set()

    async def until_stopped(self, work):
        operation = asyncio.create_task(work)
        stopping = asyncio.create_task(self.stop.wait())
        try:
            await asyncio.wait((operation, stopping), return_when=asyncio.FIRST_COMPLETED)
            if self.stopped:
                return None
            return operation.result()
        finally:
            for task in (operation, stopping):
                if not task.done():
                    task.cancel()
            await asyncio.gather(operation, stopping, return_exceptions=True)


class _RecordedWire:
    def __init__(self):
        self.entries = []
        self.followers = {}
        self.appending = False

    def head(self):
        if self.appending:
            raise AssertionError("application callback ran under append exclusion")
        return len(self.entries)

    def send(self, path: Path, message: Message):
        self.appending = True
        try:
            entry = _Entry(list(path), message, len(self.entries) + 1)
            self.entries.append(entry)
            for follower, bound in list(self.followers.items()):
                if len(follower.live.items) == bound:
                    del self.followers[follower]
                    # Signal only: the subscriber writer closes its own carrier
                    # outside append and off the producer's stack.
                    follower.end()
                else:
                    follower.live.put(entry)
        finally:
            self.appending = False

    def receive(self, receiver):
        raise WireError("no_route")

    def close(self, code=1000, reason=""):
        for follower in self.followers:
            follower.end()
        self.followers.clear()

    def attach(self, after: int, target: Wire, bound: int, pause: bool):
        follower = _Follower()
        # One synchronous critical section: append cannot enter between the
        # head snapshot and registration of its bounded live handoff.
        self.appending = True
        try:
            head = len(self.entries)
            history = self.entries[after:head]
            self.followers[follower] = bound
        finally:
            self.appending = False

        def send(entry):
            if follower.stopped:
                return False
            target.send(entry.path, entry.message)
            follower.sent.put(entry.sequence)
            return True

        async def write():
            try:
                for index, entry in enumerate(history):
                    if not send(entry):
                        return
                    if index == 0 and pause:
                        follower.paused.set()
                        await follower.until_stopped(follower.resume.wait())
                        if follower.stopped:
                            return
                while not follower.stopped:
                    entry = await follower.until_stopped(follower.live.take())
                    if entry is None or not send(entry):
                        return
            finally:
                target.close(1008, "recorded handoff ended")

        follower.done = asyncio.create_task(write())
        return head, follower


class _RecordedRoot:
    """Private root with queued asynchronous delivery, without transport APIs."""

    def __init__(self):
        self.receivers = {}
        self.queue = deque()
        self.scheduled = False
        self.closed = False

    def send(self, path: Path, message: Message):
        if self.closed:
            raise WireError("closed")
        if len(self.queue) == 16:
            raise AssertionError("witness output queue full")
        self.queue.append((list(path), message))
        if self.scheduled:
            return
        self.scheduled = True
        asyncio.get_running_loop().call_soon(self._drain)

    def _drain(self):
        self.scheduled = False
        while not self.closed and self.queue:
            path, message = self.queue.popleft()
            receiver = self.receivers.get(None)
            if receiver and receiver.message:
                receiver.message(list(path), message)

    def receive(self, receiver: Receiver):
        key = None
        if key in self.receivers:
            raise WireError("receiver_exists")
        self.receivers[key] = receiver
        active = True

        def detach():
            nonlocal active
            if active:
                active = False
                self.receivers.pop(key, None)

        return detach

    def close(self, code=1000, reason=""):
        self.closed = True
        self.queue.clear()
        self.receivers.clear()


def _message(value):
    return Message({"version": 1, "kind": "event", "data": value})


class _Presentation:
    def __init__(self, store):
        self.root = _RecordedRoot()
        self.end = _RecordedRoot()
        self.values = _Queue()
        self.closed = _Queue()
        self.failure = None
        self.root_routes = Dispatcher(self.root)
        self.end_routes = Dispatcher(self.end)
        self.destination_routes = Dispatcher(mount({"out": self.end_routes.select(["destination"])}))
        destination = self.destination_routes.select(["out"])
        self.source_routes = Dispatcher(mount({"outer": mount({"in": self.root_routes.select(["source"])})}))
        self.wire = self.source_routes.select(["outer", "in"])

        def deliver(path, message):
            try:
                if path != ["tick"]:
                    raise AssertionError("recorded destination received wrong path")
                store.head()  # Application reentry must be outside append exclusion.
                if message.frame["kind"] != "event" or type(message.frame["data"]) is not int:
                    raise AssertionError("expected event")
                self.values.put(message.frame["data"])
            except Exception as error:
                self.failure = error
                self.values.put(0)

        destination.receive(Receiver(message=deliver))

        def forward(path, message):
            try:
                destination.send(path, message)
            except Exception as error:
                self.failure = error
                self.values.put(0)

        def closed(code, reason):
            store.head()
            self.closed.put(code)

        self.wire.receive(Receiver(message=forward, closed=closed))

    async def collect(self):
        # A fence on the same route follows all earlier replay/live deliveries.
        self.wire.send(["tick"], _message(0))
        values = []
        while True:
            value = await self.values.take()
            if self.failure:
                raise self.failure
            if value == 0:
                return values
            values.append(value)

    def close(self):
        self.wire.close()
        self.source_routes.close()
        self.destination_routes.close()
        self.root_routes.close()
        self.end_routes.close()
        self.root.close()
        self.end.close()


async def _sent(follower, last):
    while await follower.sent.take() != last:
        pass  # Wait for the subscriber writer's admission barrier.


async def _head_case(before):
    store = _RecordedWire()
    first, late = _Presentation(store), _Presentation(store)
    followers = []
    try:
        source = at(mount({"record": store}), ["record"])
        for value in range(1, 4):
            source.send(["tick"], _message(value))
        if before:
            source.send(["tick"], _message(4))
        head, follower = store.attach(0, first.wire, 2, True)
        followers.append(follower)
        await follower.paused.wait()

        async def produce():
            for value in range(5 if before else 4, 6):
                source.send(["tick"], _message(value))

        # Producer completion is observed while replay remains at its barrier.
        await asyncio.create_task(produce())
        follower.resume.set()
        await _sent(follower, 5)
        source.send(["tick"], _message(6))
        await _sent(follower, 6)
        values = await first.collect()
        _, second = store.attach(3, late.wire, 2, False)
        followers.append(second)
        await _sent(second, 6)
        return {
            "cut": "append_before_head" if before else "head_before_append",
            "head": head,
            "first": values,
            "after_three": await late.collect(),
            "producer_progress": True,
            "callbacks_outside_append": True,
        }
    finally:
        store.close()
        try:
            await asyncio.gather(*(follower.done for follower in followers))
        finally:
            first.close()
            late.close()


async def _stall_case():
    store = _RecordedWire()
    stalled, healthy = _Presentation(store), _Presentation(store)
    followers = []
    try:
        for value in range(1, 4):
            store.send(["tick"], _message(value))
        _, slow = store.attach(0, stalled.wire, 2, True)
        followers.append(slow)
        await slow.paused.wait()
        _, fast = store.attach(3, healthy.wire, 2, False)
        followers.append(fast)
        for value in range(4, 6):
            store.send(["tick"], _message(value))
            await _sent(fast, value)
        queued = len(slow.live.items)
        store.send(["tick"], _message(6))
        await _sent(fast, 6)
        await asyncio.shield(slow.done)
        code = await stalled.closed.take()
        try:
            stalled.wire.send(["tick"], _message(99))
        except WireError as error:
            if error.code != "closed":
                raise
        else:
            raise AssertionError("stalled carrier accepted after close")
        underneath = _Queue()

        def probe_received(path, message):
            if message.frame["kind"] == "event":
                underneath.put(message.frame["data"])

        stalled.root_routes.register(["probe"], Receiver(message=probe_received))
        stalled.root.send(["probe"], _message(99))
        probe = await underneath.take()
        store.send(["tick"], _message(7))
        await _sent(fast, 7)
        return {
            "bound": 2,
            "queued_at_bound": queued,
            "closed": 1 + len(stalled.closed.items),
            "close_code": code,
            "healthy": await healthy.collect(),
            "underneath": [probe],
            "head": store.head(),
            "producer_progress": True,
        }
    finally:
        store.close()
        try:
            await asyncio.gather(*(follower.done for follower in followers))
        finally:
            stalled.close()
            healthy.close()


async def recorded_wire_witness(within_ms: int):
    """Exercise both head cuts and overflow isolation within one deadline."""
    async with asyncio.timeout(within_ms / 1000):
        return {"cases": [await _head_case(False), await _head_case(True)], "stalled": await _stall_case()}
